moshi-adapters-kotlinx-immutable/src/main/kotlin/dev/zacsweers/moshix/immutable/adapters/ImmutableCollectionJsonAdapter.kt

```diff
@@ -0,0 +1,158 @@
+
+    override fun create(type: Type, annotations: MutableSet<out Annotation>, moshi: Moshi): JsonAdapter<*>? {
+        if (type !is ParameterizedType) return null
+        if (annotations.isNotEmpty()) return null
+
+        return when (type.rawType) {
+            ImmutableList::class.java, ImmutableCollection::class.java,
+            PersistentList::class.java, PersistentCollection::class.java ->
+                newListAdapter<Any>(type, moshi).nullSafe()
+
+            ImmutableSet::class.java, PersistentSet::class.java ->
+                newSetAdapter<Any>(type, moshi).nullSafe()
+
+            ImmutableMap::class.java, PersistentMap::class.java ->
+                newMapAdapter(type, moshi).nullSafe()
+
+            else -> null
+        }
+    }
+
+    private fun <T> newListAdapter(type: ParameterizedType, moshi: Moshi): JsonAdapter<PersistentCollection<T?>> {
+        val elementType = Types.collectionElementType(type, PersistentCollection::class.java)
+        val elementAdapter = moshi.adapter<T>(elementType)
+        return object : ImmutableCollectionJsonAdapter<PersistentCollection<T?>, T>(elementAdapter) {
+            override fun newCollection(): PersistentCollection<T?> = persistentListOf()
+        }
+    }
+
+    private fun <T> newSetAdapter(type: ParameterizedType, moshi: Moshi): JsonAdapter<PersistentSet<T?>> {
+        val elementType = Types.collectionElementType(type, PersistentSet::class.java)
+        val elementAdapter = moshi.adapter<T>(elementType)
+        return object : ImmutableCollectionJsonAdapter<PersistentSet<T?>, T>(elementAdapter) {
+            override fun newCollection(): PersistentSet<T?> = persistentSetOf()
+        }
+    }
+
+    private fun newMapAdapter(type: ParameterizedType, moshi: Moshi): JsonAdapter<out PersistentMap<out Any, Any?>> {
+        // alternative of Types.mapKeyAndValueTypes(type, type.rawType)
+        val keyType = type.actualTypeArguments[0]
+        val valueType = type.actualTypeArguments[1]
+
+        val keyAdapter = moshi.adapter(keyType.rawType)
+        val valueAdapter = moshi.adapter(valueType.rawType)
+
+        return ImmutableMapJsonAdapter(
+            keyAdapter = keyAdapter,
+            valueAdapter = valueAdapter,
+        ).nullSafe()
+    }
+}
+
+private abstract class ImmutableCollectionJsonAdapter<E : PersistentCollection<T?>, T>(
+    private val adapter: JsonAdapter<T>
+) : JsonAdapter<E>() {
+
+    abstract fun newCollection(): E
+
+    override fun fromJson(reader: JsonReader): E {
+        val listBuilder = newCollection().builder()
+        reader.beginArray()
+        while (reader.hasNext()) {
+            listBuilder.add(
+                adapter.fromJson(reader) ?: error("[PersistentListJsonAdapter] Error parsing item: $reader")
+            )
+        }
+        reader.endArray()
+        return listBuilder.build() as E
+    }
+
+    override fun toJson(writer: JsonWriter, value: E?) {
+        writer.beginArray()
```

Comment:
If the value is null, we should emit a nullValue and return, not an empty array. That said, we also know this will never be null due to the nullSafe() use so let's just force it to non-null